pkg/services/authz/rbac/service_test.go

```diff
@@ -143,9 +149,294 @@ func TestService_checkPermission(t *testing.T) {
+			}
+
+			role, err := s.getUserBasicRole(ctx, req, userIdentifiers)
+			if tc.expectedError {
+				require.Error(t, err)
+				return
+			}
+
+			require.NoError(t, err)
+			require.Equal(t, tc.expectedRole, role)
+			if tc.cacheHit {
+				require.Zero(t, store.calls)
+			} else {
+				require.Equal(t, 1, store.calls)
+			}
+		})
+	}
+}
+
+func TestService_getUserPermissions(t *testing.T) {
+	type testCase struct {
+		name          string
+		permissions   []accesscontrol.Permission
+		cacheHit      bool
+		expectedPerms map[string]bool
+	}
+
+	testCases := []testCase{
+		{
+			name: "should return permissions from cache if available",
+			permissions: []accesscontrol.Permission{
+				{Action: "dashboards:read", Scope: "dashboards:uid:some_dashboard"},
+			},
+			cacheHit:      true,
+			expectedPerms: map[string]bool{"dashboards:uid:some_dashboard": true},
+		},
+		{
+			name: "should return permissions from store if not in cache",
+			permissions: []accesscontrol.Permission{
+				{Action: "dashboards:read", Scope: "dashboards:uid:some_dashboard"},
+			},
+			cacheHit:      false,
+			expectedPerms: map[string]bool{"dashboards:uid:some_dashboard": true},
+		},
+		{
+			name:          "should return error if store fails",
+			permissions:   nil,
+			cacheHit:      false,
+			expectedPerms: map[string]bool{},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			ctx := context.Background()
+			userID := &store.UserIdentifiers{UID: "test-uid", ID: 112}
+			req := &CheckRequest{
+				Namespace: claims.NamespaceInfo{Value: "stacks-12", OrgID: 1, StackID: 12},
+				UserUID:   userID.UID,
+				Action:    "dashboards:read",
+			}
+
+			cacheService := localcache.New(shortCacheTTL, shortCleanupInterval)
+			if tc.cacheHit {
+				cacheService.Set(userPermCacheKey(req.Namespace.Value, userID.UID, req.Action), tc.expectedPerms, 0)
+			}
+
+			store := &fakeStore{
+				userID:          userID,
+				basicRole:       &store.BasicRole{Role: "viewer", IsAdmin: false},
+				userPermissions: tc.permissions,
+			}
+
+			s := &Service{
+				store:          store,
+				identityStore:  &fakeIdentityStore{teams: []int64{1, 2}},
+				actionMapper:   mappers.NewK8sRbacMapper(),
+				logger:         log.New("test"),
+				tracer:         tracing.NewNoopTracerService(),
+				permCache:      cacheService,
+				basicRoleCache: localcache.New(longCacheTTL, longCleanupInterval),
+				teamCache:      localcache.New(shortCacheTTL, shortCleanupInterval),
+			}
+
+			perms, err := s.getUserPermissions(ctx, req)
+			require.NoError(t, err)
+			require.Len(t, perms, len(tc.expectedPerms))
+			for _, perm := range tc.permissions {
+				_, ok := tc.expectedPerms[perm.Scope]
+				require.True(t, ok)
+			}
+			if tc.cacheHit {
+				require.Equal(t, 1, store.calls) // only get user id
+			} else {
+				require.Equal(t, 3, store.calls) // get user id, basic role, and permissions
+			}
+		})
+	}
+}
+
+type fakeStore struct {
```

Comment:
nit: I've got a slight preference for having the fake store in a separate file to keep the test file a bit cleaner and allow for the fake store to be reused (unlikely in this case tbh). But also fine to keep it as it is if you like it better this way.

Reply:
Fine by me :)